### ISSUE TITLE
list index out of range on multithreaded environment

### ISSUE THREAD

**Issue by Wiston999** [elastic/elasticsearch-py#371]
Hi,
first of all, thank you very much for elasticsearch, you are doing a very nice job.
This is my first contribution to the community so please be nice with me.

I'm working on a multithreaded enviroment (using multiprocessing.pool.ThreadPool), and sometimes the following error appear:

```
File "/opt/clustering/wrappers/DAOElasticSearch/DAOElasticSearch.py", line 77, in indexDocument
    self.index(index = self.indexName, doc_type = self.docType, body = document, **kwargs)
File "/usr/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 69, in _wrapped
    return func(*args, params=params, **kwargs)
File "/usr/local/lib/python2.7/site-packages/elasticsearch/client/__init__.py", line 261, in index
    _make_path(index, doc_type, id), params=params, body=body)
File "/usr/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 326, in perform_request
    connection = self.get_connection()
File "/usr/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 182, in get_connection
    return self.connection_pool.get_connection()
File "/usr/local/lib/python2.7/site-packages/elasticsearch/connection_pool.py", line 225, in get_connection
    return self.selector.select(self.connections)
File "/usr/local/lib/python2.7/site-packages/elasticsearch/connection_pool.py", line 66, in select
    return connections[self.rr]
```

Is this a problem with my code or is this a bug of elasticsearch client?
Looking over `/usr/local/lib/python2.7/site-packages/elasticsearch/connection_pool.py`, I think that this can be solved using some kind of Lock() or Semaphore() protecting `connections` variable.

Thanks again,
Victor.

**Comment by HonzaKral** [comment]
Thanks for the report, I changed the logic to be thread safe but wasn't thorough enough. Should be fixed in master now.

**Comment by marcinn** [comment]
@HonzaKral  it seems that the issue still exists. I'm on 2.4.0 tag installed from PyPI, which contains your patch.

```
  File "/home/my-project/dwh/eggs/elasticsearch_dsl-2.2.0-py2.7.egg/elasticsearch_dsl/search.py", line 627, in execute
    **self._params
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/client/utils.py", line 69, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/client/__init__.py", line 539, in search
    doc_type, '_search'), params=params, body=body)
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/transport.py", line 324, in perform_request
    connection = self.get_connection()
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/transport.py", line 182, in get_connection
    return self.connection_pool.get_connection()
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/connection_pool.py", line 225, in get_connection
    return self.selector.select(connections)
  File "/home/my-project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/connection_pool.py", line 66, in select
    return connections[self.rr]
IndexError: list index out of range

```

The problem occurs after long run of one process which creates many threads.

**Comment by HonzaKral** [comment]
I see, the problem is also in `RoundRobinSelector` as it uses an instance variable `self.rr` without any protection against threads competing.

Thank you for the report @marcinn !

**Comment by marcinn** [comment]
Maybe storing `rr` in `threading.local()` context in `RoundRobinSelector` will resolve the issue?

**Comment by HonzaKral** [comment]
for now you can simply create your own selector class and pass it in:
```
class MySelector()
class MyRoundRobinSelector(ConnectionSelector):
    def __init__(self, opts):
        super(MyRoundRobinSelector, self).__init__(opts)
        self.data = threading.local()
        self.data.rr = -1

    def select(self, connections):
        self.data.rr += 1
        self.data.rr %= len(connections)
        return connections[self.data.rr]

es = Elasticsearch(selector_class=MyRoundRobinSelector)
```

or use `RandomSelector` as a workaround.

**Comment by marcinn** [comment]
Thanks. Most of the time I've spent searching the docs how to set custom selector_class ;)

**Comment by marcinn** [comment]
I'm getting ```AttributeError: 'thread._local' object has no attribute 'rr'```, but I'll investigate this tommorow. 

```
  File "/project/dwh/eggs/elasticsearch-2.4.0-py2.7.egg/elasticsearch/connection_pool.py", line 225, in get_connection
    return self.selector.select(connections)
  File "/project/dwh/apps/elasticroundrobin.py", line 12, in select
    self.data.rr += 1
AttributeError: 'thread._local' object has no attribute 'rr'

```

I've just checked via ipdb that `rr` is not defined for every selector instance:

```
ipdb> self; self.data.rr
<elasticroundrobin.ThreadSafeRoundRobinSelector object at 0x7f125cc55dd0>
ipdb> *** AttributeError: 'thread._local' object has no attribute 'rr'
self; self.data.rr
<elasticroundrobin.ThreadSafeRoundRobinSelector object at 0x7f125cbb0450>
-1
```

or comparing object ids and checking via hasattr:

```
ipdb> id(self),  id(self.data), hasattr(self.data, 'rr')
(139954626809488, 139954233386096, False)
ipdb> id(self),  id(self.data), hasattr(self.data, 'rr')
(139954626809488, 139954233386096, False)
ipdb> id(self),  id(self.data), hasattr(self.data, 'rr')
(139954626809488, 139954233386096, True)
ipdb> id(self),  id(self.data), hasattr(self.data, 'rr')
(139954626809488, 139954233386096, False)

```
And this is pretty strange for me as I'm pretty weak in threads programming.

**Comment by HonzaKral** [comment]
That is again my bad, I just typed it here wihout testing it, the `select` method is run in different threads whereas the `__init__` is only run once in the main thread, that explains the config.

For now please use the `RandomSelector` as a workaround.

Thank you!

**Comment by marcinn** [comment]
Hi! Yes, I'll use random selector.

I've wrapped selector creation with threadlocal (in ConnectionPool class) and I've got round robin working, but the result was similar to a random selector because every thread has own counter. To get a real round robin the `rr` incrementation should be independent from threads, I think. In the other words - there should be something like a global & thread-independent variable, if possible.

Thanks for hints and good luck!
